### PR TITLE
Add dedicated API view for A record

### PR DESCRIPTION
**Pull request by ar4s**
- [x] tests

Related to https://github.com/allegro/ralph/pull/2776

### CODE REVIEW

File: dnsaas/api/v2/views.py [hunk 1]
```diff
@@ -313,3 +315,48 @@ class TsigKeysViewSet(FiltersMixin, ModelViewSet):
...
+        data = request.data
+        assert 'action' in data
+        action = data.pop('action')
+        if action == 'delete':
+            assert 'address' in data
+            assert 'hostname' in data
+            self._delete_record(data['address'], data['hostname'])
+        else:
+            self._add_or_update_record(data)
+        return Response(data={})
```
Comment (szok) [inline dnsaas/api/v2/views.py]: maybe add "status" or something else? Not empty dict.

File: dnsaas/api/v2/views.py [hunk 2]
```diff
@@ -313,3 +315,60 @@ class TsigKeysViewSet(FiltersMixin, ModelViewSet):
...
+    def _delete_record(self, ip, hostname):
+        record = self._get_record(ip, hostname)
+        if record:
+            record.delete()
+            return 'deleted'
+        return 'noop'
+
+    def post(self, request):
+        data = request.data
+        assert 'action' in data
```
Comment (mkurek) [inline dnsaas/api/v2/views.py]: throw 400 instead

File: dnsaas/api/v2/views.py [hunk 3]
```diff
@@ -313,3 +315,60 @@ class TsigKeysViewSet(FiltersMixin, ModelViewSet):
...
+
+    def _get_record(self, ip, hostname):
+        ip = int(ipaddress.ip_address(ip))
+        try:
+            record = Record.objects.get(number=ip, name=hostname)
+        except:
+            record = None
+        return record
+
+    def _add_or_update_record(self, data):
```
Comment (mkurek) [inline dnsaas/api/v2/views.py]: what if hostname will be empty?

File: dnsaas/api/v2/views.py [hunk 4]
```diff
@@ -313,3 +315,60 @@ class TsigKeysViewSet(FiltersMixin, ModelViewSet):
...
+        new = data['new']
+        old = data['old']
+        _, created = Record.objects.update_or_create(
+            type='A',
+            number=int(ipaddress.ip_address(old['address'])),
+            name=old['hostname'],
+            domain=hostname2domain(old['hostname']),
```
Comment (mkurek) [inline dnsaas/api/v2/views.py]: what if domain will be None (will not be found using hostname)?

File: dnsaas/api/v2/tests.py [hunk 5]
```diff
@@ -846,3 +846,79 @@ def test_skips_not_specified_field(self):
...
+            'action': 'add'
+        }
+
+    def _send_post_data_to_endpoint(self):
+        return self.client.post(
+            reverse('api:v2:ip-record'), self.data, format='json',
+            **{'HTTP_ACCEPT': 'application/json; version=v2'}
+        )
+
+    def test_add_record(self):
```
Comment (ghost) [inline dnsaas/api/v2/tests.py:910]: How does it mean adding with the old values, could you describe it?

File: powerdns/models/powerdns.py [hunk 6]
```diff
@@ -345,9 +345,9 @@ class Record(TimeTrackable, Owned, RecordLike, WithRequests):
         help_text=_("The 'right hand side' of a DNS record. For an A"
                     " record, this is the IP address"),
     )
-    number = models.PositiveIntegerField(
+    number = models.DecimalField(
         _("IP number"), null=True, blank=True, default=None, editable=False,
-        db_index=True
+        db_index=True, max_digits=39, decimal_places=0
```
Comment (ghost) [inline powerdns/models/powerdns.py:350]: i'm not sure about this change on powerdns table, is this safe?

File: dnsaas/api/v2/views.py [hunk 7]
```diff
@@ -313,3 +315,60 @@ class TsigKeysViewSet(FiltersMixin, ModelViewSet):
...
+
+class IPRecordView(APIView):
+    """Dedicated view for add/update/delete"""
+    permission_classes = (IsAdminUser,)
+
+    def _get_record(self, ip, hostname):
+        ip = int(ipaddress.ip_address(ip))
+        try:
+            record = Record.objects.get(number=ip, name=hostname)
+        except:
```
Comment (ghost) [inline dnsaas/api/v2/views.py]: are you sure you want to catch all expeptions?

File: dnsaas/api/v2/views.py [hunk 8]
```diff
@@ -313,3 +315,60 @@ class TsigKeysViewSet(FiltersMixin, ModelViewSet):
...
+            status_text = self._delete_record(
+                data['address'], data['hostname']
+            )
+        else:
+            status_text = self._add_or_update_record(data)
+        return Response(
+            data={'status': status_text},
+            status=(
+                status.HTTP_200_OK
+                if status_text else status.HTTP_404_NOT_FOUND
```
Comment (mkurek) [inline dnsaas/api/v2/views.py]: when could 404 happen?

File: dnsaas/api/v2/views.py (hunk 4 above)
Comment (ghost) [reply dnsaas/api/v2/views.py]: +1 some validation at the `post` beginning could spare us a lot of troubles

File: dnsaas/api/v2/views.py (hunk 3 above)
Comment (ar4s) [reply dnsaas/api/v2/views.py]: TODO:
- if `hostname` is empty in `new` dict then delete A record
- if `hostname` is empty in `old` dict then return

File: powerdns/models/powerdns.py (hunk 6 above)
Comment (ar4s) [reply powerdns/models/powerdns.py:350]: it's custom field not powerdns

**Comment by ghost** [comment]
:ok:

**Review by szok — APPROVED** [review]

**Pull request merged**